### PR TITLE
Add sync by bittorrent snapshots

### PR DESCRIPTION
**Pull request by b00ris**
closes https://github.com/ledgerwatch/turbo-geth/issues/839

### CODE REVIEW

File: cmd/integration/commands/stages.go [hunk 1]
```diff
@@ -252,6 +264,15 @@ func stageIHash(ctx context.Context) error {
 	db := ethdb.MustOpen(chaindata)
 	defer db.Close()
 
+	if err := migrations.NewMigrator().Apply(db, ""); err != nil {
```
Comment (AlexeyAkhunov) [inline cmd/integration/commands/stages.go:267]: Do you think the second argument needs to be `datadir`? And in other places

File: cmd/state/generate/generate_header_snapshot.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	for i := uint64(1); i <= toBlock; i++ {
+		select {
+		case <-ch:
+			return errors.New("interrupted")
+		default:
+
+		}
+		hash = rawdb.ReadCanonicalHash(db, i)
+		header = rawdb.ReadHeaderRLP(db, hash, i)
+		tuples = append(tuples, []byte(dbutils.HeaderPrefix), dbutils.HeaderKey(i, hash), header)
```
Comment (AlexeyAkhunov) [inline cmd/state/generate/generate_header_snapshot.go:56]: So the header snapshot contains one out of three types of records that are in the `HeaderPrefix` table, right?

File: cmd/state/generate/generate_header_snapshot.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		log.Warn("Remove lock", "err", err)
+		return err
+	}
+	mi := metainfo.MetaInfo{
+		CreatedBy:    "turbogeth",
+		CreationDate: time.Now().Unix(),
+		Comment:      "Snapshot of headers",
+	}
+
+	info, err := torrent.BuildInfoBytesForLMDBSnapshot(snapshotPath)
```
Comment (AlexeyAkhunov) [inline cmd/state/generate/generate_header_snapshot.go:100]: Why do you do this for the headers, but not for the block bodies?

File: cmd/state/generate/remote_seeder.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+	db := ethdb.NewLMDB().Path(dir + "/tmpdb").MustOpen()
+	err := client.Run(ethdb.NewObjectDatabase(db))
+	if err != nil {
+		return err
+	}
+	torrents := client.Cli.Torrents()
+	for _, t := range torrents {
+		t := t
+		go func() {
```
Comment (AlexeyAkhunov) [inline cmd/state/generate/remote_seeder.go:39]: I suppose you could have done it with one go-routine and many timer instead (no sleeping)?

File: cmd/state/generate/seeder.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	"github.com/anacrolix/torrent/metainfo"
+	"github.com/ledgerwatch/turbo-geth/common"
+	"github.com/ledgerwatch/turbo-geth/log"
+	trnt "github.com/ledgerwatch/turbo-geth/turbo/torrent"
+	"os"
+	"os/signal"
+	"time"
+)
+
+func Seed(pathes []string) error {
```
Comment (AlexeyAkhunov) [inline cmd/state/generate/seeder.go:19]: What is the difference between this and `SeedSnapshot` in `remote_seeder` ?

File: ethdb/kv_lmdb.go [hunk 6]
```diff
@@ -681,6 +682,7 @@ func (c *LmdbCursor) initCursor() error {
 	var err error
 	c.c, err = tx.tx.OpenCursor(c.tx.db.buckets[c.bucketName].DBI)
 	if err != nil {
+		panic("su-tx" + err.Error())
```
Comment (AlexeyAkhunov) [inline ethdb/kv_lmdb.go:685]: Is it left-over or needs to be here?

File: ethdb/kv_snapshot.go [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+		forBuckets: forBucket,
+	}
+}
+
+var ErrNotSnapshotBucket = errors.New("not snapshot bucket")
+
+//lazyTx is used for lazy transaction creation.
+type lazyTx struct {
+	construct  func() (Tx, error)
+	forBuckets map[string]dbutils.BucketConfigItem
```
Comment (AlexeyAkhunov) [inline ethdb/kv_snapshot.go:226]: You seem to be only using it for the key, not for the value. Is it better to convert it to `map[string]struct{}` ?

File: ethdb/kv_snapshot.go [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+func NewSnapshotKV() snapshotOpts {
+	return snapshotOpts{
+		forBuckets: make(map[string]dbutils.BucketConfigItem),
+	}
+}
+
+type SnapshotKV struct {
+	db         KV
+	snapshotDB KV
+	forBuckets map[string]dbutils.BucketConfigItem
```
Comment (AlexeyAkhunov) [inline ethdb/kv_snapshot.go:138]: Are the values of this map actually used anywhere?

File: ethdb/kv_snapshot.go [hunk 9]
```diff
@@ -0,0 +1,502 @@
...
+}
+
+type snapshotOpts struct {
+	db         KV
+	snapshot   KV
+	forBuckets map[string]dbutils.BucketConfigItem
```
Comment (AlexeyAkhunov) [inline ethdb/kv_snapshot.go:144]: Are the values of this map used anywhere?

File: ethdb/kv_snapshot.go [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+
+	s.keyCmp = cmp
+	if cmp <= 0 {
+		return s.lastDBKey, s.lastDBVal, nil
+	}
+
+	return s.lastSNDBKey, s.lastSNDBVal, nil
+}
+
+func (s *snapshotCursor) SeekTo(seek []byte) ([]byte, []byte, error) {
```
Comment (AlexeyAkhunov) [inline ethdb/kv_snapshot.go:422]: Hasn't `seekTo` been deleted?

**Review by crypt1d — COMMENTED** [review]
Rather silly thing, there seems to be a small typo in the filenames:

`cmd/state/commands/bodies_snashot.go`
`cmd/state/commands/headers_snashot.go`

File: cmd/state/generate/generate_header_snapshot.go (hunk 2 above)
Comment (b00ris) [reply cmd/state/generate/generate_header_snapshot.go:56]: Yes, you are right. Just list of canonical headers. We can build header number to header hash/td and canonical indexes after snapshot download.

**Comment by mandrigin** [comment]
@b00ris just a silly question: is there any command-line flag or something to limit sync to be able to seed, say exactly state at the block 10_000_000?

**Comment by b00ris** [comment]
@mandrigin Do you mean: I want to use snapshots for sync but I want to start stage sync from block N?

File: cmd/state/generate/generate_header_snapshot.go (hunk 3 above)
Comment (b00ris) [reply cmd/state/generate/generate_header_snapshot.go:100]: I used it just to see snapshot hash

File: cmd/state/generate/seeder.go (hunk 5 above)
Comment (b00ris) [reply cmd/state/generate/seeder.go:19]: SeedSnapshot is for start seeding from scratch. remote_seeder can be used to download snapshots and continue seeding.

File: cmd/integration/commands/stages.go (hunk 1 above)
Comment (b00ris) [reply cmd/integration/commands/stages.go]: fixed

**Comment by mandrigin** [comment]
@b00ris if I want to seed a snapshot, I maybe want some nice round number to seed.

**Comment by b00ris** [comment]
@mandrigin No, you can seed whole database file. It can be prepared by generate_headers/bodies snapshot command.

**Pull request merged**